### PR TITLE
Use configured realm and shard for system files web3 module

### LINKED ISSUE #10602: Use configured realm and shard for system files in web3
### Problem

When interacting with system files, the web3 API needs to ensure the configured realm and shard are used.

### Solution

* Update `SystemFileLoader` to pass configured shard and realm to `FileID` construction
* Update `RatesAndFeesLoader`
* Update `hedera.mirror.web3.evm.fundingAccount`
* Update `ServicesConfiguration` `systemAccountDetector` and `strictSystemAccountDetector`
* Update `PrecompilesHolder` to put both zero realm/shard precompile entry and current realm/shard (TBD)

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by filev94**
**Description**:
This PR modifies several classes in web3 module to use configured shard and realm from common properties.
Changes include:
    - Modify SystemFileLoader and RatesAndFeesLoader to include CommonProperties and set the shard and realm in the class from them.
    - Validate fundingAccount in MirrorNodeEvmProperties - if the correct account is not set based on realm and shard, the flow creates the correct one and sets it.
    - Modify systemAccountDetectors to extract the entityId from the address and compare the num with the boundary
    - Changes in the EntityRepository queries that fetch based on alias or address to include the shard and realm, as well as changes in the corresponding classes and tests
   
**Related issue(s)**:

Fixes #10602 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/EntityDatabaseAccessor.java [hunk 1]
```diff
@@ -30,7 +32,7 @@ public class EntityDatabaseAccessor extends DatabaseAccessor<Object, Entity> {
             if (isMirror(addressBytes)) {
                 return getEntityByMirrorAddressAndTimestamp(address, timestamp);
             } else {
-                return getEntityByEvmAddressAndTimestamp(addressBytes, timestamp);
+                return getEntityByEvmAddressTimestampShardAndRealm(addressBytes, timestamp);
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/EntityDatabaseAccessor.java:35]: This method name doesn't need to change as it's implementation should be encapsulated from caller.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java [hunk 2]
```diff
@@ -190,6 +195,12 @@ public class MirrorNodeEvmProperties implements EvmProperties {
     @DecimalMax("1.0")
     private double modularizedTrafficPercent = 0.0;
 
+    @Autowired
+    public MirrorNodeEvmProperties(CommonProperties commonProperties) {
+        this.commonProperties = commonProperties;
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:200]: Why does it need to be stored if it's only used in the constructor?

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java [hunk 3]
```diff
@@ -190,6 +195,12 @@ public class MirrorNodeEvmProperties implements EvmProperties {
     @DecimalMax("1.0")
     private double modularizedTrafficPercent = 0.0;
 
+    @Autowired
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:198]: This autowired shouldn't be needed as spring should add it implicitly for single constructor.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java [hunk 4]
```diff
@@ -89,8 +95,7 @@ public class MirrorNodeEvmProperties implements EvmProperties {
     @DurationMin(seconds = 1)
     private Duration expirationCacheTime = Duration.ofMinutes(10L);
 
-    @NotBlank
-    private String fundingAccount = "0x0000000000000000000000000000000000000062";
+    private EntityId fundingAccount;
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:98]: This is a breaking property change. Revert to storing as hex string and convert default entityid to hex string in constructor.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/EntityRepository.java [hunk 5]
```diff
@@ -23,7 +23,7 @@ public interface EntityRepository extends CrudRepository<Entity, Long> {
             cacheManager = CACHE_MANAGER_ENTITY,
             key = "T(java.util.Arrays).hashCode(#alias)",
             unless = "#result == null")
-    Optional<Entity> findByEvmAddressAndDeletedIsFalse(byte[] alias);
+    Optional<Entity> findByEvmAddressAndDeletedIsFalseAndShardAndRealm(byte[] alias, long shard, long realm);
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/EntityRepository.java:26]: For all changed repo methods, prefer shard and realm to be the first parameters instead of the last. Thus also prefer they be ?1 and ?2 in sql as well.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/SystemFileLoader.java [hunk 6]
```diff
@@ -49,7 +51,11 @@ private Map<FileID, File> loadAll() {
     }
 
     private File load(int fileNum, Bytes contents) {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/SystemFileLoader.java:53]: We should add entries for each system file in `SystemEntity` then use that here. This avoids having the hardcoded numbers above and have a central way to trace these files in the codebase.

```suggestion
    private File load(SystemEntity systemFile, Bytes contents) {
```

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/CommonEntityAccessor.java [hunk 7]
```diff
@@ -44,10 +48,12 @@ public class CommonEntityAccessor {
                 .orElseGet(() -> entityRepository.findByIdAndDeletedIsFalse(entityId.getId()));
     }
 
-    public Optional<Entity> getEntityByEvmAddressAndTimestamp(
+    public Optional<Entity> getEntityByEvmAddressTimestampShardAndRealm(
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/CommonEntityAccessor.java:51]: Revert method rename

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/entity/SystemEntity.java [hunk 8]
```diff
@@ -14,7 +14,9 @@ public enum SystemEntity {
     FEE_COLLECTOR_ACCOUNT(98L),
     NODE_REWARD_ACCOUNT(801L),
     STAKING_REWARD_ACCOUNT(800L),
-    TREASURY_ACCOUNT(2L);
+    TREASURY_ACCOUNT(2L),
+    EXCHANGE_RATE(112L),
+    FEE_SCHEDULE(111L);
```
[the remark this answers is not in the document]
Comment (filev94) [reply hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/entity/SystemEntity.java]: Addressed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/EntityDatabaseAccessor.java (hunk 1 above)
Comment (filev94) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/EntityDatabaseAccessor.java]: Reverted back to previous

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java (hunk 2 above)
Comment (filev94) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java]: Removed it, i was only testing something and forgot to delete it.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java (hunk 3 above)
Comment (filev94) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:195]: When i remove the @Autowired annotation the integration tests start failing with failed to load context

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java (hunk 4 above)
Comment (filev94) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java]: Reverted back to hex String and setting the value in the constructor

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/EntityRepository.java (hunk 5 above)
Comment (filev94) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/EntityRepository.java]: Addressed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/SystemFileLoader.java (hunk 6 above)
Comment (filev94) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/SystemFileLoader.java]: Addressed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/CommonEntityAccessor.java (hunk 7 above)
Comment (filev94) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/CommonEntityAccessor.java]: Addressed

File: hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/entity/SystemEntity.java [hunk 9]
```diff
@@ -11,7 +11,16 @@
...
     ADDRESS_BOOK_102(102L),
+    EXCHANGE_RATE(112L),
     FEE_COLLECTOR_ACCOUNT(98L),
+    FEE_SCHEDULE(111L),
+    FILE_101(101L),
+    FILE_102(102L),
+    FILE_111(111L),
+    FILE_112(112L),
+    FILE_121(121L),
+    FILE_122(122L),
```
Comment (steven-sheehy) [inline hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/entity/SystemEntity.java:22]: Not sure why we're duplicating numbers and why we're not giving them descriptive names like the other entries.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/evm/pricing/RatesAndFeesLoaderTest.java [hunk 10]
```diff
@@ -168,12 +174,15 @@ void loadEmptyFeeSchedules() {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/evm/pricing/RatesAndFeesLoaderTest.java:174]: We should add a test with non-zero realm and shard

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/evm/pricing/RatesAndFeesLoaderTest.java [hunk 11]
```diff
@@ -66,18 +68,18 @@ class RatesAndFeesLoaderTest {
             .consensusTimestamp(300L)
             .fileData("corrupt".getBytes())
             .build();
-    private static final String CORRUPT_RATES_MESSAGE = "Rates 0.0.112 are corrupt!";
-    private static final long EXCHANGE_RATES_ID = 112L;
-    private static final String CORRUPT_SCHEDULES_MESSAGE = "Fee schedule 0.0.111 is corrupt!";
-    private static final long FEE_SCHEDULES_ID = 111L;
-    private static final long EXCHANGE_RATE_ID = 112L;
+    private static final long EXCHANGE_RATES_ID = SystemEntity.EXCHANGE_RATE.getNum();
+    private static final long FEE_SCHEDULES_ID = SystemEntity.FEE_SCHEDULE.getNum();
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/evm/pricing/RatesAndFeesLoaderTest.java:72]: We shouldn't use the num as it won't work with non-zero realm/shard.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/state/SystemFileLoaderTest.java [hunk 12]
```diff
@@ -13,15 +13,17 @@
...
 import com.hedera.hapi.node.transaction.ExchangeRateSet;
 import com.hedera.hapi.node.transaction.ThrottleDefinitions;
+import com.hedera.mirror.common.CommonProperties;
 import com.hedera.mirror.web3.evm.properties.MirrorNodeEvmProperties;
 import java.time.Instant;
 import java.time.temporal.ChronoUnit;
 import org.assertj.core.api.InstanceOfAssertFactories;
 import org.junit.jupiter.api.Test;
 
 class SystemFileLoaderTest {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/state/SystemFileLoaderTest.java:23]: We should have a test with non-zero realm and shard

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:1]: Can you please also set `HederaConfig` shard and realm props in `MirrorNodeEvmProperties.buildTransactionProperties()`?

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/evm/pricing/RatesAndFeesLoaderTest.java (hunk 11 above)
Comment (filev94) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/evm/pricing/RatesAndFeesLoaderTest.java]: Refactored to use getId()

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/state/SystemFileLoaderTest.java (hunk 12 above)
Comment (filev94) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/state/SystemFileLoaderTest.java:29]: Addressed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/evm/pricing/RatesAndFeesLoaderTest.java (hunk 10 above)
Comment (filev94) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/evm/pricing/RatesAndFeesLoaderTest.java:205]: Addressed

File: hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/entity/SystemEntity.java (hunk 9 above)
Comment (filev94) [reply hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/entity/SystemEntity.java]: Removed duplicates and renamed

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by jnels124 — APPROVED** [review]
